src/main/kotlin/io/sip3/salto/ce/recording/RecordingHandler.kt

```diff
@@ -0,0 +1,196 @@
+        config().getJsonObject("recording")?.let { config ->
+            config.getLong("trim-to-size-delay")?.let {
+                trimToSizeDelay = it
+            }
+            config.getLong("expiration-delay")?.let {
+                expirationDelay = it
+            }
+            config.getLong("aggregation-timeout")?.let {
+                aggregationTimeout = it
+            }
+            config.getInteger("bulk-size")?.let {
+                bulkSize = it
+            }
+        }
+
+        vertx.setPeriodic(trimToSizeDelay) {
+            bulks = MutableMapUtil.mutableMapOf(bulks)
+        }
+        vertx.setPeriodic(expirationDelay) {
+            val now = System.currentTimeMillis()
+            bulks.filterValues { it.createdAt + aggregationTimeout < now }
+                .forEach { (callId, recordingRow) ->
+                    writeToDatabase(recordingRow)
+                    bulks.remove(callId)
+                }
+        }
+
+        vertx.eventBus().localConsumer<Packet>(RoutesCE.rec) { event ->
+            try {
+                val packet = event.body()
+                handle(packet)
+            } catch (e: Exception) {
+                logger.error("RecordingHandler 'handle()' failed.", e)
+            }
+        }
+
+        GlobalScope.launch(vertx.dispatcher()) {
+            val index = vertx.sharedData().getLocalCounter(RoutesCE.rec).await()
+            vertx.eventBus()
+                .localConsumer<Pair<Packet, RecordingPayload>>( RoutesCE.rec + "_${index.andIncrement.await()}") { event ->
+                    try {
+                        val (packet, recording) = event.body()
+                        handleRecording(packet, recording)
+                    } catch (e: Exception) {
+                        logger.error(e) { "RecordingHandler 'handleRecording()' failed." }
+                    }
+                }
+        }
+    }
+
+    open fun handle(packet: Packet) {
+        val recording = RecordingPayload().apply {
+            val payload = Unpooled.wrappedBuffer(packet.payload)
+            decode(payload)
+        }
+
+        if (recording.type == PacketTypes.RTCP) {
+            val rtcpPacket = Packet().apply {
+                timestamp = packet.timestamp
+                srcAddr = packet.srcAddr
+                dstAddr = packet.dstAddr
+                protocolCode = PacketTypes.RTCP
+                source = "sip3"
+                payload = recording.payload
+            }
+            vertx.eventBus().localSend(RoutesCE.rtcp, rtcpPacket)
+        }
+
+        val index = abs(recording.callId.hashCode()) % instances
+        vertx.eventBus().localSend(RoutesCE.rec + "_${index}", Pair(packet, recording))
+    }
+
+    open fun handleRecording(packet: Packet, recording: RecordingPayload) {
+        val key = "${recording.callId}:${packet.srcAddr.compositeKey(packet.dstAddr)}"
+        val recordingRow = bulks.getOrPut(key) {
+            RecordingRow(packet.createdAt).apply {
```

Comment:
Do we need a parameter for constructor? It seems that assigning all the value explicitly is just fine